src/systems/physics/Physics.cc

```diff
@@ -1723,209 +1723,274 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm)
-            _ecm.SetChanged(_entity, components::Pose::typeId,
-                ComponentState::PeriodicChange);
-          }
-        }
-        IGN_PROFILE_END();
+  IGN_PROFILE_BEGIN("Models");
+  // Model entities and their world poses.
+  std::unordered_map<Entity, math::Pose3d> modelWorldPoses;
 
-        // Populate world poses, velocities and accelerations of the link. For
-        // now these components are updated only if another system has created
-        // the corresponding component on the entity.
-        auto worldPoseComp = _ecm.Component<components::WorldPose>(_entity);
-        if (worldPoseComp)
+  _ecm.Each<components::Model, components::Pose, components::ParentEntity,
+            components::Static, components::ModelCanonicalLink>(
+      [&](const Entity &_entity, components::Model *, components::Pose *_pose,
+          components::ParentEntity *_parentEntity, components::Static *_static,
+          components::ModelCanonicalLink *_canonicalLink) -> bool
+      {
+        if (_static->Data())
         {
-          auto state =
-              worldPoseComp->SetData(math::eigen3::convert(frameData.pose),
-              this->pose3Eql) ?
-              ComponentState::PeriodicChange :
-              ComponentState::NoChange;
-          _ecm.SetChanged(_entity, components::WorldPose::typeId, state);
+          return true;
         }
 
-        // Velocity in world coordinates
-        auto worldLinVelComp =
-            _ecm.Component<components::WorldLinearVelocity>(_entity);
-        if (worldLinVelComp)
+        auto linkFrameIt = linkFrameData.find(_canonicalLink->Data());
+        if (linkFrameIt == linkFrameData.end())
         {
-          auto state = worldLinVelComp->SetData(
-                math::eigen3::convert(frameData.linearVelocity),
-                this->vec3Eql) ?
-                ComponentState::PeriodicChange :
-                ComponentState::NoChange;
-          _ecm.SetChanged(_entity,
-              components::WorldLinearVelocity::typeId, state);
+          ignerr << "Internal error: Frame data for link [" << _entity
+                 << "] not found" << std::endl;
+          return true;
         }
-
-        // Angular velocity in world frame coordinates
-        auto worldAngVelComp =
-            _ecm.Component<components::WorldAngularVelocity>(_entity);
-        if (worldAngVelComp)
-        {
-          auto state = worldAngVelComp->SetData(
-              math::eigen3::convert(frameData.angularVelocity),
-              this->vec3Eql) ?
-              ComponentState::PeriodicChange :
-              ComponentState::NoChange;
-          _ecm.SetChanged(_entity,
-              components::WorldAngularVelocity::typeId, state);
+        std::optional<math::Pose3d> parentWorldPose;
+
+        // If this model is nested, we assume the pose of the parent model has
+        // already been updated. We expect to find the updated pose in
```

Comment:
This assumption is currently valid because `ecm.Each` uses `Views` which hold their entities in a `std::set` (ordered set): https://github.com/ignitionrobotics/ign-gazebo/blob/78caaffb4469c6072b1cdd5315f2deb8444cf4ab/include/ignition/gazebo/detail/View.hh#L121

The `Entity` values are generated in ascending order during entity creation and the ordered set preserves that order. If that ever changes to a `std::unordered_set`, this assumption will not hold anymore.

However, this is all very implicit and I don't know if the `std::set` was used for this purpose. The alternative is to keep a topologically sorted list of models in the physics system. If we want to go that route, I would say, let's do that in a follow up PR.

Reply:
Yeah it's tricky to depend on an implementation detail like that. I think it's fine for now, but we should consider making that an explicit feature that we support in the long run.

Reply:
Thanks @azeey for looking into this. I think that creating an alternative data structure in a future PR is probably a good idea. I went ahead and created an issue for this so that we can re-visit it later: https://github.com/ignitionrobotics/ign-gazebo/issues/706

Reply:
Thank you for creating the issue!